### ISSUE TITLE
Sentinel-1 and Sentinel-2 image alignment for Self-Supervised Segmentation pre-training

### ISSUE THREAD

**Issue by KIselborn** [zhu-xlab/SSL4EO-S12#22]
Hello, I intended to use SSL4EO-S12 for multimodal learning and therefore was inspecting the provided sample images in QGIS. However, whilst doing so we observed that the images seem to be misaligned when using the given coordinates. For example in the image with ID `0000002` it looks as if the Sentinel-2 image is contained within the Sentinel-1 (with an offset of 2 to 3 pixels on each side), instead of covering the same area on the ground:
<img width="1916" alt="image" src="https://github.com/zhu-xlab/SSL4EO-S12/assets/105886192/a2bf413a-43ce-41a1-8401-4a45eb9658c5">
For other images such as the one with ID `0000005`, the images seem to be slightly shifted:
![image](https://github.com/zhu-xlab/SSL4EO-S12/assets/105886192/a8c56523-e9f4-466a-9b2d-3385aab5f297)
Also note that for the image with ID `0000002` the coordinates seem to define a weird rectangular instead of square area...

The Problem is now, that in order to use both modalities for a combined segmentation model, one would of course require a precise alignment of the raster-imagery. Unfortunately, I cannot tell whether the coordinates set in the tif-files or the actually extracted images are the problem here. In the former case, this would just wind down to a visual bug that can be ignored during training, however in the latter case either interpolation to a common raster (requiring knowledge of the ground-truth bounding box coordinates) or re-downloading would be necessary as far as I can tell.

Looking at the [download-code](https://github.com/zhu-xlab/SSL4EO-S12/blob/d511fec91933b4debd0713fac2ae8c496a218d1b/src/download_data/ssl4eo_downloader.py#L284), it seems like the images are (after being fetched from the google earth engine) extracted using a simple center-crop and with the bounding coordinates being computed using linear interpolation based on the image size in EPSG:4326 [If I'm not mistaken, these coordinates are later applied to the tif-files].  (I'm not familiar with GEE, so I would assume that S1 + S2 images are automatically being projected as well to this coordinate system instead of staying in the local coordinate system they are provided in for download, so that I guess that this is not a Problem?) Whilst Sentinel-1 and Sentinel-2 may not have exactly matching ground-resolutions, with the difference only noticable when downloading very large portions of the raw data (e.g. ~10 pixels when downloading images of size around 12k pixels in Germany), this doesn't really explain this significantly larger offset.  With these confusing observations, could you please clarify the logic behind the bounding-coordinates calculation and the alignment of the Sentinel-1 and Sentinel-2 images? Is it really correct to just feed them in a stacked manner into a multimodal model?

I would appreciate any help that can be provided with this issue, so that I can employ this dataset for my work. It is a great combination of spatially  and temporally distributed multimodal data - thanks for all the effort of collecting, hosting, etc. the dataset and the ongoing effort in answering Questions 😃

**Comment by wangyi111** [comment]
Hi! Thanks for your very detailed inspection and feedback!

Regarding the logic behind the alignment, like you mentioned, as I know GEE stores raw Sentinel tiles in their local CRS and the re-projection is done online during downloading. The projection itself should not be an issue. However, I think there're some tiny misalignment afterwards. First, when I sampled a 264x264 rectangle around the center coords, the resulting patch size could be slightly different (e.g. 263x266 for s1 and 265x264 for s2). After center cropping and resizing to 264x264, they are already not perfectly aligned, but as the example above should be only 1 or 2 pixels offset. This means if you stack s1 and s2 rasters, there can exist such small offsets. Second, to store a GeoTiff file for each patch, we matched the bounding boxes of the raw rectangle to the newly center-cropped patch. This possibly results in the shift you observed in QGIS, as the pixel coords are not exactly the same as in real world.

With the above, I think the "precise" pixel-by-pixel alignment can not be ensured in this dataset unfortunately (can have offset for 1 or 2 pixels, but not as big as the second example you show in QGIS). However, I would assume a simple stacking will not cause a very big issue (just my thinking:)). I've stacked them to train multimodal self-supervised algorithms and it worked fine. For segmentation, if the intended mask is not required to be very precise for each pixel, it should be rather okish. For sentinel level I think it's usually hard even for experts to label very accurately. Of course it depends on your specific goal:)

Hope I answered your question. Let me know if anything else is not clear!

**Comment by KIselborn** [comment]
Hi, thanks for your detailed response - yes this answers my question :).
We appreciate your insights and thank you for confirming that there are some issues w.r.t. to the image alignment. Since we are considering tasks which are very sensitive to even small offsets, we are unsure about how applicable this dataset then is for our target objective. Especially since further comparisons with e.g. OpenStreetMap show offsets like in the image with ID`0000005` indicating that the coordinates are not that precise.

Note that for example rivers can be only a single pixel wide in a Sentinel-1/2 image...

**Comment by wangyi111** [comment]
I see, in that case one needs to be careful indeed.
